signer/src/utxo.rs

```diff
@@ -0,0 +1,964 @@
+    /// the given request. Only a valid signature is needed to satisfy the
+    /// deposit script.
+    fn as_tx_input(&self, sig: Signature) -> TxIn {
+        TxIn {
+            previous_output: self.outpoint,
+            script_sig: ScriptBuf::new(),
+            sequence: Sequence(0),
+            witness: self.construct_witness_data(sig),
+        }
+    }
+
+    /// Construct the witness data for the taproot script of the deposit.
+    ///
+    /// Deposit UTXOs are taproot spend what a "null" key spend path,
+    /// a deposit script-path spend, and a redeem script-path spend. This
+    /// function creates the witness data for the deposit script-path
+    /// spend where the script takes only one piece of data as input, the
+    /// signature. The deposit script is:
+    ///
+    ///   <data> OP_DROP OP_DUP OP_HASH160 <pubkey_hash> OP_EQUALVERIFY OP_CHECKSIG
+    ///
+    /// where <data> is the stacks deposit address and <pubkey_hash> is
+    /// given by self.deposit_public_key. The public key used for key-path
+    /// spending is self.taproot_public_key, and is supposed to be a dummy
+    /// public key.
+    pub fn construct_witness_data(&self, sig: Signature) -> Witness {
+        let (internal_key, _) = self.taproot_public_key.inner.x_only_public_key();
+        let ver = LeafVersion::TapScript;
+
+        // For such a simple tree, we construct it by hand.
+        let leaf1 = NodeInfo::new_leaf_with_ver(self.deposit_script.clone(), ver);
+        let leaf2 = NodeInfo::new_leaf_with_ver(self.redeem_script.clone(), ver);
+
+        // A Result::Err is returned by NodeInfo::combine if the depth of
+        // our taproot tree exceeds the maximum depth of taproot trees,
+        // which is 128. We have two nodes so the depth is 1 so this will
+        // never panic.
+        let node =
+            NodeInfo::combine(leaf1, leaf2).expect("This tree depth greater than max of 128");
+        let taproot = TaprootSpendInfo::from_node_info(SECP256K1, internal_key, node);
+
+        // TaprootSpendInfo::control_block returns None if the key given,
+        // (script, version), is not in the tree. But this key is definitely
+        // in the tree (see the variable leaf1 above).
+        let control_block = taproot
+            .control_block(&(self.deposit_script.clone(), ver))
+            .expect("We just inserted the deposit script into the tree");
+
+        let witness_data = [
+            sig.serialize().to_vec(),
+            self.deposit_script.to_bytes(),
+            control_block.serialize(),
+        ];
+        Witness::from_slice(&witness_data)
+    }
+}
+
+#[derive(Debug)]
+pub struct WithdrawalRequest {
+    /// The amount of BTC, in sats, to withdraw.
+    pub amount: u64,
+    /// The max fee amount to use for the sBTC deposit transaction.
+    pub max_fee: u64,
+    /// The public key that will be able to spend the output.
+    pub public_key: PublicKey,
```

Comment:
That all makes sense to me, but I think that would lead us being off the low-level designs in the [Deposit API LLD](https://docs.google.com/document/d/13AN9GdLfPYKKPkeHNU4niq_CyJCDxFR7DGBLzLc5gxU/edit#heading=h.vwfrlp5xn7lz) and the [Clarity LLD](https://docs.google.com/document/d/1Bwi1jM_7p9OqNGT41eWCuoKnfoOrNHycqxby3XSonCA/edit#heading=h.y6v41emwi920). I was following the Deposit API LLD, which uses the public key, while the clarity LLD uses a tuple of a 1 byte `version` field with a 32 byte `hashbytes` field (which I think is a compressed public key).

I'm okay changing the designs and it's all equally easy for me to update here. Just want to call out that this would lead to a mismatch.